From 084f52c3044b6a1c0d12deb764542bc9de6a886f Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Wed, 26 Oct 2022 14:12:14 +0800
Subject: [PATCH] sw64: improve stack trace
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 084f52c3044b6a1c0d12deb764542bc9de6a886f
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I645NW

--------------------------------

Process ra directly if call stack has not been setup yet so we won't
miss the previous function.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>
Reviewed-by: He Sheng <hesheng@wxiat.com>
Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kernel/stacktrace.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/sw_64/kernel/stacktrace.c b/arch/sw_64/kernel/stacktrace.c
index 7b5ddc78bd6d..70c142b11b92 100644
--- a/arch/sw_64/kernel/stacktrace.c
+++ b/arch/sw_64/kernel/stacktrace.c
@@ -69,8 +69,18 @@ void walk_stackframe(struct task_struct *tsk, struct pt_regs *regs,
 	struct stackframe frame;
 
 	if (regs) {
+		unsigned long offset;
 		pc = regs->pc;
 		fp = regs->r15;
+		if (kallsyms_lookup_size_offset(pc, NULL, &offset)
+				&& offset < 16) {
+			/* call stack has not been setup
+			 * store pc first then loop from ra
+			 */
+			if (fn(pc, data))
+				return;
+			pc = regs->r26;
+		}
 	} else if (tsk == current || tsk == NULL) {
 		fp = (unsigned long)__builtin_frame_address(0);
 		pc = (unsigned long)walk_stackframe;
-- 
2.33.0

